src/freenet/node/MultiMessageCallback.java

```diff
@@ -2,14 +2,37 @@
 
 import freenet.io.comm.AsyncMessageCallback;
 
-/** Waits for multiple asynchronous message sends, then calls finish(). */
+/** Waits for multiple asynchronous message sends, then calls finish(). 
+ * You should add messages with make() and then call arm(). sent(boolean)
+ * will be called when all the messages have been sent (or failed e.g. 
+ * disconnected) and finish(boolean) will be called when all the messages 
+ * have been acknowledged (or failed). Typical usage:
+ * <pre>Message m1 = ...;
+ * Message m2 = ...;
+ * PeerNode pn = ...;
+ * MultiMessageCallback mcb = new MultiMessageCallback() {
+ *   protected void finish(boolean success) {
+ *     // Messages have finished.
+ *   }
+ *   protected void sent(boolean success) {
+ *     // Messages have been sent.
+ *   }
+ * }
+ * pn.sendAsync(m1, mcb, ctr);
+ * pn.sendAsync(m2, mcb, ctr);
+ * mcb.arm();</pre> */
 public abstract class MultiMessageCallback {
 	
+    /** Number of messages that have not yet completed */
 	private int waiting;
+	/** Number of messages that have not yet been sent */
 	private int waitingForSend;
 	
+	/** True if arm() has been called. finish(boolean) and sent(boolean) will 
+	 * only be called after arming the callback. */
 	private boolean armed;
```

Comment:
A group of packets. The whole point here is we get a callback when we've sent all of a bunch of messages. And then another callback when they're all "finished" (i.e. acknowledged). But not before we call arm(), for the sake of consistency of the caller. See the example in the top comment.